src/main/scala/is/hail/methods/UpperIndexBounds.scala

```diff
@@ -0,0 +1,65 @@
+package is.hail.methods
+
+import is.hail.HailContext
+import is.hail.annotations.UnsafeIndexedSeq
+import is.hail.expr.types._
+import is.hail.linalg.BlockMatrix
+import is.hail.table.Table
+import is.hail.utils.fatal
+import org.apache.spark.sql.Row
+
+object UpperIndexBounds {
+
+  def groupPositionsByContig(tbl: Table): Array[Array[Int]] = {
+    val expectedSignature = TStruct("contig" -> TString(), "pos" -> TInt32())
+    if (tbl.signature != expectedSignature) {
+      fatal(s"Expected table to have signature $expectedSignature, but found ${ tbl.signature }.")
+    }
+    
+    val groupedPositions = tbl.keyBy("contig").groupByKey("positions").collect()
```

Comment:
I'd parallelize sorting and computeUpperIndexBounds by mapping the underlying RDD before collecting.

In a future PR, we could consider adding computeUpperIndexBounds as a Hail function. Then it can be applied more directly to the grouped Table to yield another Table, and also exposed in Python.

Reply:
Patrick, here we'd like to groupByKey on a Table with one value field of type T to get an Array[T] field per key rather than an Array[Struct[T]] field per key. How does this operation relate to your PR?
https://github.com/hail-is/hail/pull/3345/files

Reply:
map the struct after using collect_by_key?

Reply:
I'd forgotten about collect_by_key on Table, that's the ticket.